market/mk20/ddo_v1.go

```diff
@@ -0,0 +1,191 @@
+package mk20
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/ethereum/go-ethereum"
+	eabi "github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/yugabyte/pgx/v5"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/builtin/v16/verifreg"
+
+	"github.com/filecoin-project/curio/harmony/harmonydb"
+)
+
+var UnknowContract = errors.New("provider does not work with this market")
+
+// DDOV1 defines a structure for handling provider, client, and piece manager information with associated contract and notification details
+// for a DDO deal handling.
+type DDOV1 struct {
+
+	// Provider specifies the address of the provider
+	Provider address.Address `json:"provider"`
+
+	// Client represents the address of the deal client
+	Client address.Address `json:"client"`
+
+	// Actor able to with AuthorizeMessage (like f1/f3 wallet) able to authorize actions such as managing ACLs
+	PieceManager address.Address `json:"piecemanager"`
+
+	// Duration represents the deal duration in epochs. This value is ignored for the deal with allocationID.
+	// It must be at least 518400
+	Duration abi.ChainEpoch `json:"duration"`
+
+	// AllocationId represents an aggregated allocation identifier for the deal.
+	AllocationId *verifreg.AllocationId `json:"aggregatedallocationid"`
+
+	// ContractAddress specifies the address of the contract governing the deal
+	ContractAddress string `json:"contractaddress"`
+
+	// ContractDealIDMethod specifies the method name to retrieve the deal ID for a contract
+	ContractDealIDMethod string `json:"contractdealidmethod"`
```

Comment:
The deal ID method is also where we do the first off-chain verification? Personally I'd call that method 'Verify' as duplicate id check is really just part of verification